lib/collection/src/wal_delta.rs

```diff
@@ -1172,4 +1192,314 @@ mod tests {
         );
         assert_eq!(resolve_result.unwrap_err(), WalDeltaError::NotFound);
     }
+
+    /// Assert that we `check_clock_tag_ordering_property` on the WAL.
+    async fn assert_wal_ordering_property(wal: &RecoverableWal) {
+        // Grab list of clock tags from WAL records, skip non-existent or below cutoff tags
+        let clock_tags = {
+            let cutoff = wal.cutoff_clocks.lock().await;
+            wal.wal
+                .lock()
+                .read(0)
+                // Only take records with clock tags
+                .filter_map(|(_, operation)| operation.clock_tag)
+                // Clock tags must be equal or higher to cutoff point
+                .filter(|clock_tag| {
+                    cutoff
+                        .get_tick(clock_tag.peer_id, clock_tag.clock_id)
+                        .map_or(true, |a| clock_tag.clock_tick >= a)
+                })
+                .collect::<Vec<_>>()
+        };
+
+        check_clock_tag_ordering_property(&clock_tags).expect("WAL ordering property violated");
+    }
+
+    /// Test that we satisfy the clock ordering property, allowing WAL recovery resolution.
+    ///
+    /// Property:
+    /// For each operation with peer+clock tick X, all following operations having the same
+    /// peer+clock must cover ticks X+1, X+2, ..., X+n in order up to the highest tick value of
+    /// that peer+clock in the WAL.
+    ///
+    /// More specifically, this tests the property again on every clock tag. The result of this
+    /// check is that the sequence always ends in order at the end, going up to the highest clock
+    /// clock tick.
+    ///
+    /// This logic is validated with examples in `validate_clock_tag_ordering_property`.
+    ///
+    /// This property may not be valid if a diff transfer has not been resolved correctly or
+    /// completely, or if the WAL got malformed in another way.
+    fn check_clock_tag_ordering_property(clock_tags: &[ClockTag]) -> Result<(), String> {
+        // Get the highest clock value for each clock+peer
+        let highest_clocks = clock_tags
+            .iter()
+            .fold(HashMap::new(), |mut map, clock_tag| {
+                map.entry((clock_tag.peer_id, clock_tag.clock_id))
+                    .and_modify(|highest| *highest = clock_tag.clock_tick.max(*highest))
+                    .or_insert(clock_tag.clock_tick);
+                map
+            });
```

Comment:
🪡 Just use a loop, man. Embrace the caveman, don't fall for func-prog lobby propaganda. We have real loops in Rust, we don't need to `fold` stuff...